nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-administration/src/main/java/org/apache/nifi/admin/service/EntityStoreAuditService.java

```diff
@@ -0,0 +1,636 @@
+     *
+     * @param directory Persistent Entity Store directory
+     */
+    public EntityStoreAuditService(final File directory) {
+        environment = loadEnvironment(directory);
+        entityStore = PersistentEntityStores.newInstance(environment);
+        logger.info("Environment configured with directory [{}]", directory);
+    }
+
+    /**
+     * Add Actions to Persistent Store
+     *
+     * @param actions Collections of Actions to be added
+     */
+    @Override
+    public void addActions(final Collection<Action> actions) {
+        Objects.requireNonNull(actions, "Actions required");
+
+        entityStore.executeInExclusiveTransaction(storeTransaction -> {
+            for (final Action action : actions) {
+                addAction(storeTransaction, action);
+            }
+            logger.debug("Actions added [{}]", actions.size());
+        });
+    }
+
+    /**
+     * Get Previous Values for specified Component Identifier
+     *
+     * @param componentId Component Identifier for which previous property values should be retrieved
+     * @return Map of Property Name to List of Previous Property Values
+     */
+    @Override
+    public Map<String, List<PreviousValue>> getPreviousValues(final String componentId) {
+        Objects.requireNonNull(componentId, "Component Identifier required");
+
+        return entityStore.computeInReadonlyTransaction(storeTransaction -> {
+            final Map<String, List<PreviousValue>> previousValuesFound = new LinkedHashMap<>();
+
+            final EntityIterable actionEntities = storeTransaction.find(EntityType.ACTION.getEntityType(), ActionEntity.SOURCE_ID.getProperty(), componentId);
+            for (Entity actionEntity : actionEntities) {
+                final Entity configureDetails = actionEntity.getLink(ActionLink.CONFIGURE_DETAILS.getProperty());
+                if (configureDetails != null) {
+                    final String name = getProperty(configureDetails, ConfigureDetailsEntity.NAME);
+                    final String value = getProperty(configureDetails, ConfigureDetailsEntity.VALUE);
+
+                    final PreviousValue previousValue = new PreviousValue();
+                    previousValue.setPreviousValue(value);
+                    previousValue.setUserIdentity(getProperty(actionEntity, ActionEntity.USER_IDENTITY));
+                    previousValue.setTimestamp(getDateProperty(actionEntity, ActionEntity.TIMESTAMP));
+
+                    final List<PreviousValue> previousValues = previousValuesFound.get(name);
+                    if (previousValues == null) {
+                        final List<PreviousValue> newPreviousValues = new ArrayList<>();
+                        newPreviousValues.add(previousValue);
+                        previousValuesFound.put(name, newPreviousValues);
+                    } else if (previousValues.size() < PREVIOUS_VALUES_LIMIT) {
+                        previousValues.add(previousValue);
+                    }
```

Comment:
nitpick and not worth delaying the PR, but if there are other problems that arise, it's worth nothing that this can be simplified to just:
```
final List<PreviousValue> previousValues = previousValuesFound.computeIfAbsent(name, () -> new ArrayList<>());
previousValuesFound.add(previousValue);
```